### PR TITLE
Add branching model team agreement doc

### LINKED ISSUE #14: Branching Model Team Agreement
A place to start on a few branching model team agreements to consider adopting:

- Be as lightweight as possible in our approach, and only focus on adopting team agreements that will add value to our [future selves](https://xkcd.com/1597/).
- Adopt the `main` branch as our [default branch](https://github.com/button-inc/digital_marketplace/settings/branches).
- Agree never to commit directly on `main` under any circumstances, and implement [branch protection rules](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/defining-the-mergeability-of-pull-requests/about-protected-branches) as a guardrail to enforce the agreement, including enforcement to administrators of this repo.
- Make draft pull requests early and often to facilitate a transparent process.
- Ruthlessly delete branches that aren't tied directly to an open pull request.
- Adopt [Conventional Commits](https://www.conventionalcommits.org/en/) standard for commit messages, and consider authoring commit messages using [Commitizen](http://commitizen.github.io/cz-cli/) to make the spec easier while ensuring notes [stay informative](https://xkcd.com/1296/).
- Follow [peer review best practices](https://google.github.io/eng-practices/review/reviewer/standard.html) by suggesting opportunities to improve code during peer review, merging as soon as the code is *better* than the code in the target branch and [release ready](https://trunkbaseddevelopment.com/observed-habits/#always-release-ready).
- Treat any opportunity for improvement feedback identified during peer review but not implemented in the PR where it was raised as technical debt worthy of a new issue referencing the PR where the comments first came up.
- Commit finalized agreements in `docs/branching_model.md` as a living document, to ensure any decision can be updated as and when required.

### PR DESCRIPTION
**Pull request by mikevespi**
Starting point for the agreement discussed by the team.

### CODE REVIEW

File: docs/branching_model_team_agreement.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+- Pull requests must pass CI checks in order to be merged to `main`
+- Branches are automatically deleted when merged
+- Squash merge has been disabled
+
+## Practices
+- Make draft pull requests early and often to facilitate a transparent process
+- Reviews can be done by the first person who gets to it. If the code needs explaining, request the author to walk you through it
+- Follow peer review best practices by suggesting opportunities to improve code during peer review, merging as soon as the code is better than the code in the target branch and release ready
+- Treat any opportunity for improvement feedback identified during peer review but not implemented in the PR where it was raised as technical debt worthy of a new issue referencing the PR where the comments first came up
+- Branch naming policy is as follows: dm-issue#-brief-description
```
Comment (kriscooke) [inline docs/branching_model_team_agreement.md]: Is the `dm-` prefix performing some meaningful function? (per the example of this branch being `dm-14-branching-model-agreement`) Or could we do with a more succinct `issue#-brief-description`?

**Review by kriscooke — DISMISSED** [review]

File: docs/branching_model_team_agreement.md (hunk 1 above)
Comment (wenzowski) [reply docs/branching_model_team_agreement.md]: In the upstream there's a pattern (maybe team agreement) of beginning each branch with the JIRA ticket number. Perhaps we continued that practice in this repo? A team agreement on consistent naming would be helpful...though there's often a one-to-many relation between an issue and the PRs that close it. Perhaps we could consider including the ticket number we're closing in the description section of the commit that substantially delivers the feature? For single-commit pull requests (or draft PRs that begin after pushing the very first commit) the title and description of the commit message are auto populated into the github PR title and description.

File: docs/branching_model_team_agreement.md [hunk 2]
```diff
@@ -0,0 +1,18 @@
+# Branching Model Team Agreement
+
+This document lists the agreed upon practices the team will be using. As they are discussed and changed, update this file accordingly.
+
+## Guardrails
+- The default branch for this repo is `main`
+- all commits must be made to a non-protected branch and submitted via a pull request before they can be merged into `main`. (There can be no direct commits made to `main`.)
+- Pull requests require at least 1 approval before the can be merged to `main`. Any commits made to an approved branch will dismiss the approval. (If you commit new code to a branch after it gets approved but before you merge it, you have to get it approved again.)
```
Comment (wenzowski) [inline docs/branching_model_team_agreement.md:8]:
```suggestion
- Pull requests require at least 1 approval before they can be merged to `main`. Any commits made to an approved branch will dismiss the approval. (If you commit new code to a branch after it gets approved but before you merge it, you have to get it approved again.)
```

File: docs/branching_model_team_agreement.md [hunk 3]
```diff
@@ -0,0 +1,18 @@
+# Branching Model Team Agreement
+
+This document lists the agreed upon practices the team will be using. As they are discussed and changed, update this file accordingly.
+
+## Guardrails
+- The default branch for this repo is `main`
+- all commits must be made to a non-protected branch and submitted via a pull request before they can be merged into `main`. (There can be no direct commits made to `main`.)
```
Comment (wenzowski) [inline docs/branching_model_team_agreement.md:7]:
```suggestion
- All commits must be made to a non-protected branch and submitted via a pull request before they can be merged into `main`. (There can be no direct commits made to `main`.)
```

File: docs/branching_model_team_agreement.md [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+- Pull requests require at least 1 approval before the can be merged to `main`. Any commits made to an approved branch will dismiss the approval. (If you commit new code to a branch after it gets approved but before you merge it, you have to get it approved again.)
+- Pull requests must pass CI checks in order to be merged to `main`
+- Branches are automatically deleted when merged
+- Squash merge has been disabled
```
Comment (wenzowski) [inline docs/branching_model_team_agreement.md:11]:
```suggestion
- Squash merge has been disabled to increase efficiency of `git bisect`
```

File: docs/branching_model_team_agreement.md [hunk 5]
```diff
@@ -0,0 +1,18 @@
+# Branching Model Team Agreement
+
+This document lists the agreed upon practices the team will be using. As they are discussed and changed, update this file accordingly.
+
+## Guardrails
+- The default branch for this repo is `main`
+- all commits must be made to a non-protected branch and submitted via a pull request before they can be merged into `main`. (There can be no direct commits made to `main`.)
+- Pull requests require at least 1 approval before the can be merged to `main`. Any commits made to an approved branch will dismiss the approval. (If you commit new code to a branch after it gets approved but before you merge it, you have to get it approved again.)
+- Pull requests must pass CI checks in order to be merged to `main`
+- Branches are automatically deleted when merged
```
Comment (wenzowski) [inline docs/branching_model_team_agreement.md:10]:
```suggestion
- Branches are automatically deleted on Github when a PR is merged. Each teammate should set `git config fetch.prune true` to mirror this behaviour to their local workstation.
```

**Review by wenzowski — COMMENTED** [review]
Maybe clarity on why the team agreed to these merging guardrails helps understand each guardrail?

File: docs/branching_model_team_agreement.md (hunk 1 above)
Comment (BCerki) [reply docs/branching_model_team_agreement.md]: The `dm-` prefix is something we're doing in the upstream, so we decided to keep it the same here

Comment (wenzowski) [reply docs/branching_model_team_agreement.md]:
```suggestion
- Branch naming policy is as follows: brief-description
```

Comment (kriscooke) [inline docs/branching_model_team_agreement.md]:
```suggestion
- Branch naming policy is as follows: issue#-brief-description
```

**Review by kriscooke — DISMISSED** [review]
Let's get started with this, and can follow up on some of the editorial revisions suggested here with #21

File: docs/branching_model_team_agreement.md (hunk 1 above)
Comment (wenzowski) [reply docs/branching_model_team_agreement.md]: Perhaps there's a risk that by keeping it the same as upstream we will confuse references to JIRA ticket numbers (the DM-xx form) with the Github Issues of this fork? In standup @BCerki asked if the suggestion above was a typo. The option of beginning branches with the issue number they relate to `issue#-brief-description` is another possible team agreement, and the impact of that decision would be to create a mechanism whereby we avoid making pull requests that don't correspond to tracked issues. As we were discussing in standup, sometimes a quick refactor is best enabled by directly sending a pull request without creating an issue–sometimes trivial refactors unblock our feature tickets. Additionally, I'd note that our branches are a transitory artifact–per suggestion above migrating the issue linkage to commit messages themselves would make this ephemeral linkage permanent and immutable.

**Review by kriscooke — APPROVED** [review]

File: docs/branching_model_team_agreement.md (hunk 1 above)
Comment (wenzowski) [reply docs/branching_model_team_agreement.md]: Marked this unresolved until we fully migrate discussion to #21 so it shows when people follow link from that issue.

**Pull request merged**